GreengrassIPC.md

```diff
@@ -0,0 +1,172 @@
+`connect` previously succeeded; connection establishment failures only manifest as a rejected promise.
+
+Calling `close` on a connected client will cause a `disconnection` event to be emitted as part of the shutdown process.
+
+### Client Error Handling
+All errors thrown from the Greengrass IPC implementation are of the 
+[RpcError](https://aws.github.io/aws-iot-device-sdk-js-v2/node//classes/eventstream_rpc.RpcError.html) type.  The RpcError class 
+contains properties for the type of error, a description, an optional, potentially-recursive inner error, and an 
+optional modeled error.  Modeled errors represent failures sent by the Greengrass nucleus.
+
+All client input is heavily validated for conformance to the Greengrass RPC service contract.  Some validation (like enum
+values) is skipped to allow for model evolution while retaining backwards compatibility.
+
+## Greengrass IPC Operations
+Once connected, a client can be used to perform IPC operations.
+Greengrass IPC operations fall into two categories: request-response and streaming.
+
+### Request-Response Operations
+All IPC operations that conform to a single request and single response use the standard asynchronous request-response
+pattern.  Invoking these operations is a simple matter of await-ing on the client API call:
+
+```typescript
+await client.publishToIotCore({
+    topicName: "hello/world",
+    payload: "Hello from a component!",
+    qos : greengrasscoreipc.model.QOS.AT_LEAST_ONCE    
+});
+```
+
+If a request-response operation fails, the associated promise will be rejected with an 
+[RpcError](https://aws.github.io/aws-iot-device-sdk-js-v2/node//classes/eventstream_rpc.RpcError.html) value that 
+contains more information about what went wrong.
+
+### Streaming Operations
+While streaming operations can in theory be bi-directional, the Greengrass RPC service model currently only contains
+operations with streaming output.
+
+A streaming operation is not performed automatically when the client API is invoked (otherwise the user would not have
+an opportunity to attach event listeners beforehand, potentially missing streaming events).  Instead the client API
+creates and returns a `StreamingOperation` object.
+
+A `StreamingOperation` object supports several events:
+* `message` - event emitted every time a message is successfully received on the stream from the server.  The data included with this event is specific to each streaming operation type.
+* `streamError` - event emitted when either the deserialization of an incoming message failed, or the deserialization resulted in a modeled service error.  The data included with this event is an [RpcError](https://aws.github.io/aws-iot-device-sdk-js-v2/node//classes/eventstream_rpc.RpcError.html).
+* `ended` - event emitted when the stream has been closed for any reason.  Only emitted for operations that were successfully activated.
```

Comment:
Nitpick: add `prior to stream close` after `activated`.

Reply:
You can't activate a stream after close so it may be a little redundant